File: advcake/stockmann/get_product.py
import re
import requests
import random
from bs4 import BeautifulSoup
from get_sale import get_sale
from get_brand import get_brand
from get_sizes import get_sizes
from get_prices import get_prices
from get_images import get_images
from get_categories import get_categories
from get_transliterate import get_transliterate
from get_info_color_structure_country import get_info_color_structure_country


def get_product(html, link):
    try:
        soup = BeautifulSoup(html, 'lxml')
        # получаю крошки для распаршивания категорий, возраста и пола
        age_cat_subcat = get_categories(soup)
        if age_cat_subcat:
            age = age_cat_subcat['age']
            category = age_cat_subcat['category']
            gender = age_cat_subcat['gender']
            subcategory = age_cat_subcat['subcategory']
        else: return
        # получаю цены
        price = get_prices(soup.find('li', class_=re.compile('prices_discount__')))
        oldprice = get_prices(soup.find('li', class_=re.compile('prices_price__')))
        sale = get_sale(soup)
        # получаю инфо, цвет, состав и страну & season
        info_color_structure_country = get_info_color_structure_country(soup)
        info = info_color_structure_country['info']
        color = info_color_structure_country['color']
        country = info_color_structure_country['country']
        season = info_color_structure_country['season']
        structure = info_color_structure_country['structure']
        # получаю бренд
        brand = get_brand(soup)['brand']
        name = get_brand(soup)['name']
        # получаю размеры в наличии
        if category == 'Аксессуары': sizes = ['one size']
        else: sizes = get_sizes(soup, brand, category)
        # получаю пикчи
        images = get_images(soup)
        if len(sizes) == 0 or len(sizes) == 0: return
        if category == 'Верхняя одежда ': category = 'Одежда'
        return {
            'id': round(random.uniform(1000000000, 9999999999)),
            'age': age,
            'benefit': oldprice - price,
            'brand': brand,
            "brandCountry": False,
            'brandCountry_t': False,
            'category': category,
            'category_t': get_transliterate(category),
            'color': color,
            'color_t': get_transliterate(color) if color else color,
            'country': country,
            'country_t': get_transliterate(country) if country else country,
            'delivery': ['ru'],
            'deliveryPrice': 199,
            'description': False,
            'gender': gender,
            'name': name,
            'info': info,
            'installment': False,
            'images': images,
            'like': 0,
            'link': link,
            'oldprice': int(oldprice),
            'pp': 'advcake',
            'price': int(price),
            'sale': int(sale),
            'season': season,
            'season_t': get_transliterate(season) if season else season,
            'shop': 'stockmann',
            'sizes': sizes,
            "style": False,
            'style_t': False,
            'structure': structure,
            'subcategory': subcategory,
            'subcategory_t': get_transliterate(subcategory)
        }
    except:
        print(f'{link} НЕ собран!')
        return


# res = requests.get('https://stockmann.ru/product/5187316-dzhinsy-mavi/?oid=ctl2etk3u&utm_source=partners&utm_medium=cpa&utm_campaign=335&utm_content=gc9g7&wid=gc9g7&statid=351_77bda930feaa5ad5&sub5=48832&clickId=77bda930feaa5ad53146a5d7e2dfdaeb&sub1=77bda930feaa5ad5&sub2=3146a5d7e2dfdaeb&sub=77bda930feaa5ad5')
# print(get_product(res.text, 'link'))

File: advcake/stockmann/getProduct.js
const axios = require('axios')
const getOptions = require('./getOptions')
const getTransliterate = require('./getTransliterate')


module.exports = async function getProduct (productProto) {
  try {
    // получаю данные по товару
    const productOptions = getOptions(null, productProto.link)
    const { data } = await axios.get( ...productOptions )

    if (!data || !data.payload) return
    productProto.colors.forEach(productColor => colorValue = productColor.xmlId === productProto.xmlId && productColor.name !== '' ? productColor.name : false)
    data.payload.properties.forEach(productSeason => seasonValue = productSeason.name === 'Сезон' ? productSeason.value : false)
    data.payload.properties.forEach(productStyle => styleValue = productStyle.name === 'Стиль' ? productStyle.value : false)

    const sizesArray = []
    data.payload.sizes.forEach(productSize => productSize.online === true
      ? productSize.second !== ''
        ? sizesArray.push(productSize.second.split(/[\/-]+/))
        : sizesArray.push(productSize.main.split(/[\/-]+/))
      : false)

    const notUniqueSizes = []
    sizesArray.forEach(size => {
      notUniqueSizes.push(...size)
    })

    const sizeConversion = {
      3:34.5,3.5:35,4:35.5,4.5:36,5:36.5,5.5:37,6:37.5,6.5:38,7:39,7.5:40,8:40.5,8.5:41,
      9:42,9.5:43,10:43.5,10.5:44,11:45,11.5:46,12:46.5,12.5:47,13:47.5,13.5:48,14:48.5,
      14.5:49,15:50,15.5:51,16:52,16.5:53,17:53.5,17.5:54,18:55
    }
    const convertSize = usSize => sizeConversion[+usSize.replace(',', '.')] || usSize
    const replaceOneSize = size => size === 'один размер' || size === 'ONESI' || size === 'ONESIZE'
      || size === 'ONE SIZE' || size === 'ONE' || size === 'NR' || size === 'NOSIZE' || size === '1'
      || size.includes(' см') || size.includes(' см')
      || size === 'OS' || size === 'OSO' || size === 'NO SIZE' || size === 'x' || size === '000'
      ? 'one size'
      : size

    const uniqueSizes = [ ...new Set(notUniqueSizes)].map(convertSize).map(replaceOneSize)
    uniqueSizes.forEach(function (size, ind) {
      uniqueSizes[ind] = size.replaceAll('XXXXL', '4XL').replaceAll('XXXL', '3XL').replaceAll('XXL', '2XL')
    })

    const product = {
      id: Math.floor(Math.random() * 9e9) + 1e9,
      age: 'Взрослый',
      benefit: productProto.price - productProto.priceDiscount,
      brand: productProto.brand.toUpperCase(),
      brandCountry: false,
      brandCountry_t: false,
      category: data.payload.breadcrumbs[2].name === 'Инвентарь'
        ? 'Аксессуары'
        : data.payload.breadcrumbs[2].name,
      category_t: data.payload.breadcrumbs[2].name === 'Инвентарь'
        ? 'Аксессуары'
        : getTransliterate(data.payload.breadcrumbs[2].name),
      color: colorValue && !colorValue.includes(' ')
      ? colorValue.replaceAll('ё', 'е').replaceAll(';', '-')
      : false,
      color_t: colorValue && !colorValue.includes(' ')
        ? getTransliterate(colorValue.replaceAll('ё', 'е').replaceAll(';', '-'))
        : false,
      country: false,
      country_t: false,
      create: new Date,
      delivery: ['ru'],
      deliveryPrice: 199,
      description: data.payload.description,
      gender: data.payload.breadcrumbs[1].name === "Мужчинам"
        ? 'Мужской'
        : 'Женский',
      installment: false,
      images: productProto.images.map(image => image.default.webp.src2x),
      like: 0,
      link: 'https://stockmann.ru' + productProto.link,
      name: productProto.name + ' ' + productProto.brand,
      shop: 'stockmann',
      info: data.payload.properties.length >= 3
        ? data.payload.properties.slice(2).map(property => property.name + ': ' + property.value)
        : false,
      oldprice: productProto.price,
      pp: 'advcake',
      price: productProto.priceDiscount,
      sale: productProto.discount,
      season: seasonValue,
      season_t: seasonValue
        ? getTransliterate(seasonValue)
        : false,
      sizes: uniqueSizes[0] === 'один размер'
        ? ['one size']
        : uniqueSizes,
      style: styleValue,
      style_t: styleValue
        ? getTransliterate(styleValue)
        : false,
      structure: data.payload.properties[1].value.replaceAll(';', ',').split(','),
      subcategory: data.payload.breadcrumbs[data.payload.breadcrumbs.length - 3].name,
      subcategory_t: getTransliterate(data.payload.breadcrumbs[data.payload.breadcrumbs.length - 3].name).replaceAll(' ', '-')
    }
    if (product.subcategory === 'Сумки' || product.subcategory === 'Рюкзаки'
      || product.subcategory === 'Бейсболки' || product.subcategory === 'Гетры'
      || product.subcategory === 'Козырьки' || product.subcategory === 'Кошельки'
      || product.subcategory === 'Мешки для обуви' || product.subcategory === 'Напульсники'
      || product.subcategory === 'Панамы' || product.subcategory === 'Плавки'
      || product.subcategory === 'Повязки' || product.subcategory === 'Шапки'
      || product.subcategory === 'Шарфы' || product.subcategory === 'Перчатки') {
      product.category = 'Аксессуары'
      product.category_t = getTransliterate(product.category)
    }
    if (product.subcategory === 'Весенние' || product.subcategory === 'Классические' || product.subcategory === 'Аксессуары'
      || product.subcategory === 'Мужчинам' || product.subcategory === 'Женщинам' || product.subcategory === 'Одежда'
      || product.subcategory === 'Обувь') return
    if (product.sizes.length === 0) return
    if (product.category.length > 20) return
    if (product.category.includes('чин') || product.category.includes('щин') || product.category.includes('г')
      || product.category.includes('л') || product.category.includes('нн')
      || product.category.includes('з') || product.category.includes('до')) return
    return product
  } catch (e) {
    console.log('Товар', 'https://stockmann.ru' + productProto.link, 'не удалось собрать!')
    return null
  }
}
